# 예외처리

## 1. 예외의 개념
### 예외(exception)
- 프로그램 실행 도중에 발생할 수 있는 비정상적인 사건
    - 비정상적인 데이터, 자원의 부족 등
- 예외 상황에 대한 적절한 대비를 하지 않으면 프로그램이 안정적으로 실행되지 않는 문제가 일어날 수 있음
- 예외처리(exception handling)
    - 프로그램 실행 중 예외가 발생하였을 때를 대비하여 마련해 놓은 처리 절차에 따라 대응하는 것

### 예외가 발생하는 상황
- 정상적인 처리를 할 수 없는 데이터
```C++
double hmean(double a, double b) // 조화평균
{
    // a == -b인 경우 실행 중 프로그램 비정상 종료
    if (a == -b) {
        cout << "나누기를 할 수 없습니다." << endl;
        exit(EXIT_FAILURE); // 프로그램 강제 종료
    }
    return 2.0 * a * b / (a + b);
}
```
- 프로그램이 요청하는 자원을 할당할 수 없는 경우
```C++
void f() {
    int *p = new(nothrow) int[1000000000];
    // 메모리 할당에 실패한 경우 실행 중 에러가 발생하여 프로그램이 비정상적으로 종료함
    // 메모리 할당이 이루어지지 않으면 new(nothrow)는 nullptr를 반환하므로, 이를 이용하여 메모리 할당 오류를 검사함
    if (!p) { // p가 nullptr이면
        cerr << "메모리 할당 오류" << endl;
        exit(EXIT_FAILURE);
    }
}
```


## 2. C++ 언어의 예외처리 체계
### C++ 언어의 예외처리 구문
- try 블록, catch 블록, 그리고 throw 문장으로 구성
```C++
RetType1 someFunction()
{
    try {
        // 예외가 발생할 수 있는 부분
        someDangerousFunction();
    }
    catch (eClass e) {
        // 발생한 예외를 처리하는 부분
        exceptionProcRtn();
    }
}

RetType2 someDangerousFunction()
{
    if (예외검출조건)
        throw eObj; // 예외 발생을 알림 
    else
        // 정상적인 처리
}
```

### 예외 유형별 처리
- 하나의 try 블록에 여러 개의 catch 블록 사용
- throw된 예외 객체의 자료형에 맞는 매개변수가 선언된 catch 블록에서 예외를 처리함
```C++
try {
    // 예외가 발생할 가능성이 있는 함수 호출
}
catch (eClass1 e) {
    // 예외처리 블록1
}
catch (eClass2 e) {
    // 예외처리 블록2
}
```

### 예외처리에 따른 자원 관리 문제
- 자원 소실이 가능한 상황 
```C++
void f()
{
    int *p = new int[1000];
    for (int i = 0; i < 1000; i++)
        p[i] = i;
    if (ex_condition)
        throw "exception";
    
    delete[] p;
}
```
    -> 예외를 처리할 catch 블록으로 복귀할 때 f()가 호출될 때까지 거쳐온 함수들의 지역변수들은 정상적인 소멸 과정을 거침

### 스마트 포인터의 활용
- unique_ptr: 할당된 메모리를 한 개의 포인터만 가리킬 수 있음
    - 다른 unique_ptr에 대입할 수 없으며, 이동 대입만 할 수 있음
    - unique_ptr가 제거되거나 nullptr를 대입하면 가리키고 있던 메모리를 반납함
- shared_ptr: 할당된 메모리를 여러 개의 포인터로 가리킬 수 있음
    - 다른 shared_ptr에 대입 및 이동 대입 가능
    - 포인터가 제거되거나 nullptr을 대입하는 등의 처리로 그 메모리를 가리키는 shared_ptr이 더 이상 없으면 메모리를 반납함
```C++
// unique_ptr 활용
#include <memory>
void f()
{
    std::unique_ptr<int[]> p { new int[1000] };
    for (int i = 0; i < 1000; i++)
        p[i] = i;
    if (ex_condition)
        throw "exception";
}

// vector 활용
void f()
{
    std::vector<int> p(1000);
    for (int i = 0; i < 1000; i++)
        p[i] = i;
    if (ex_condition)
        throw "exception";
}
```

### noexcept 지정자
- 함수가 예외를 일으키지 않음을 지정
```C++
template <typename T>
T max(const vector<T>& v) noexcept {
    auto p = v.begin();
    T m = *p++;
    for (; p != v.end(); p++)
        if (m < *p) m = *p;
    return m;
}
```


## 3. 예외처리 클래스
### 클래스에서 예외처리 활용하기
- 예외처리 클래스의 활용
    - 클래스 설계 시 예외처리 기능을 포함시킴으로써 객체에서 예외가 발생하였을 때 그 위치나 원인 등의 식별을 용이하게 할 수 있음
    - 클래스 선언문 내에 예외처리 담당 클래스를 선언하여 활용함

### 예외 객체의 멤버를 통한 예외 정보 전달 
```C++
// IntArray1.h
class Array {
    ......
public:
    ......
    class BadIndex {
    public:
        int wrongIndex;
        BadIndex(int n) : wrongIndex(n) {}
    };
};
// IntArray1.cpp
int& Array::operator[](int offset)
{
    if (offset < 0 || offset >= size) // 예외조건 검사
        throw BadIndex(offset);
    return buf[offset];
}
// IA1Main.cpp
int main()
{
    ......
    try {
        for (int i = 0; i <= 10; i++)
            arr[i] = i;
    }
    catch (Array::BadIndex e) {
        cerr << "인덱스 범위 오류 --> " << e.worngIndex << endl;
    }
};

```


### exception 클래스
- C++ 언어에서 예외를 처리하기 위해 예외처리 담당 클래스의 기초 클래스로 제공하는 클래스
- 헤더파일 <exception>을 소스파일에 포함시킴
- 가상함수 what()을 멤버함수로 가지고 있음
    - 예외의 종류를 char* 형태로 반환함
    - exception의 파생 클래스에서 재정의하여 사용함

### 예외 객체의 다시 던지기
- catch 블록에서 처리를 완결할 수 없는 예외의 전달
    - 현 단계의 catch 블록에서 처리를 완결할 수 없는 예외에 대한 후속 처리를 할 수 있게 예외 객체를 다시 throw할 수 있음 